File: Cpp/2711_difference_of_number_of_distinct_values_on_diagonals.cpp
// https://leetcode.cn/problems/difference-of-number-of-distinct-values-on-diagonals/

#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;
        
class Solution {
public:
    vector<vector<int>> differenceOfDistinctValues(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        vector<vector<int>> res(m, vector<int>(n, 0));

        auto update = [&](int row, int col) {
            unordered_map<int, int> topleft, bottomright;
            int i = row, j = col;
            while (i < m && j < n) {
                bottomright[grid[i][j]] += 1;
                i += 1;
                j += 1;
            }
            i = row;
            j = col;
            while (i < m && j < n) {
                int val = grid[i][j];
                bottomright[val] -= 1;
                if (bottomright[val] <= 0) bottomright.erase(val);

                res[i][j] = abs((int)topleft.size() - (int)bottomright.size());

                topleft[val] += 1;
                i += 1;
                j += 1;
            }
        };

        for (int row = 0; row < m; ++row) update(row, 0);
        for (int col = 1; col < n; ++col) update(0, col);
        return res;
    }
};

int main() {
    Solution sol;
    vector<vector<int>> grid = {{1,2,3}, {3,1,5}, {3,2,1}};
    auto res = sol.differenceOfDistinctValues(grid);
    for (auto &row: res) {
        for (auto val: row) cout << val << ", ";
        cout << '\n';
    }
    return 0;
}

File: Python3/2711_difference_of_number_of_distinct_values_on_diagonals.py
# https://leetcode.cn/problems/difference-of-number-of-distinct-values-on-diagonals/

from collections import defaultdict

class Solution:
    def differenceOfDistinctValues(self, grid):
        """
        Inputs:
            grid: list[list[int]]
        Outputs:
            res: list[list[int]]
        """
        def update(row, col):
            topleft, bottomright = defaultdict(int), defaultdict(int)
            i, j = row, col
            while i < m and j < n:
                bottomright[grid[i][j]] += 1
                i += 1
                j += 1
            i, j = row, col
            while i < m and j < n:
                val = grid[i][j]
                bottomright[val] -= 1
                if bottomright[val] <= 0:
                    del bottomright[val]

                res[i][j] = abs(len(topleft) - len(bottomright))

                topleft[val] += 1
                i += 1
                j += 1

        m, n = len(grid), len(grid[0])
        res = [[0 for _ in range(n)] for _ in range(m)]
        for row in range(m):
            update(row, 0)
        for col in range(1, n):
            update(0, col)
        return res